# Python Linear Algebra_Part 6: Integer Programming & Relaxation
## Full Day Workshop for user learn Data Science with Python
## 2018 Timothy CL Lam
This is meant for internal usage, part of contents copied externally, not for commercial purpose`

# NP-Completeness
- So far we've seen a lot of good news: such-and-such a problem can be solved quickly (in close to linear time, or at least a time that is some small polynomial function of the input size).
- NP-completeness is a form of bad news: evidence that many important problems can't be solved quickly.


- Use a heuristic. If you can't quickly solve the problem with a good worst case time, maybe you can come up with a method for solving a reasonable fraction of the common cases.
- Solve the problem approximately instead of exactly. A lot of the time it is possible to come up with a provably fast algorithm, that doesn't solve the problem exactly but comes up with a solution you can prove is close to right.
- Use an exponential time solution anyway. If you really have to solve the problem exactly, you can settle down to writing an exponential time algorithm and stop worrying about finding a better solution.
- Choose a better abstraction. The NP-complete abstract problem you're trying to solve presumably comes from ignoring some of the seemingly unimportant details of a more complicated real world problem. 


- Classification of problems

**P**. 
- Problems that can be solved in polynomial time. 
- ("P" stands for polynomial.) 


**NP**. 
- This stands for "nondeterministic polynomial time" where nondeterministic is just a fancy way of talking about guessing a solution. 
- A problem is in NP if you can quickly (in polynomial time) test whether a solution is correct 
- NP does not stand for "non-polynomial". There are many complexity classes that are much harder than NP.

### So P or NP
- Although defined theoretically, many of these classes have practical implications. 
- For instance **P** is a very good approximation to the class of problems which can be solved quickly in practice -- 
- usually if this is true, we can prove a polynomial worst case time bound, and conversely the polynomial time bounds we can prove are usually small enough that the corresponding algorithms really are practical. 
- NP-completeness theory is concerned with the distinction between the first two classes, P and NP.

### Examples:

#### Example 1: Long simple paths.
A simple path in a graph is just one without any repeated edges or vertices. To describe the problem of finding long paths in terms of complexity theory, we need to formalize it as a yes-or-no question: given a graph G, vertices s and t, and a number k, does there exist a simple path from s to t with at least k edges? A solution to this problem would then consist of such a path.

Why is this in NP? If you're given a path, you can quickly look at it and add up the length, double-checking that it really is a path with length at least k. This can all be done in linear time, so certainly it can be done in polynomial time.

However we don't know whether this problem is in P; I haven't told you a good way for finding such a path (with time polynomial in m,n, and K). And in fact this problem is NP-complete, so we believe that no such algorithm exists.


#### Example 2: Cryptography.

Suppose we have an encryption function e.g.

    code=RSA(key,text)
    
The "RSA" encryption works by performing some simple integer arithmetic on the code and the key, which consists of a pair (p,q) of large prime numbers. One can perform the encryption only knowing the product pq; but to decrypt the code you instead need to know a different product, (p-1)(q-1).
A standard assumption in cryptography is the "known plaintext attack": we have the code for some message, and we know (or can guess) the text of that message. We want to use that information to discover the key, so we can decrypt other messages sent using the same key.

Formalized as an NP problem, we simply want to find a key for which code=RSA(key,text). If you're given a key, you can test it by doing the encryption yourself, so this is in NP.

The hard question is, how do you find the key? For the code to be strong we hope it isn't possible to do much better than a brute force search.

One can test a factorization quickly (just multiply the factors back together again), so the problem is in NP. Finding a factorization seems to be difficult, and we think it may not be in P. However there is some strong evidence that it is not NP-complete either; it seems to be one of the (very rare) examples of problems between P and NP-complete in difficulty.

#### Example 3: Chess.

We've seen in the news recently a match between the world chess champion, Gary Kasparov, and a very fast chess computer, Deep Blue. The computer lost the match, but won one game and tied others.

What is involved in chess programming? Essentially the sequences of possible moves form a tree: The first player has a choice of 20 different moves (most of which are not very good), after each of which the second player has a choice of many responses, and so on. Chess playing programs work by traversing this tree finding what the possible consequences would be of each different move.

The tree of moves is not very deep -- a typical chess game might last 40 moves, and it is rare for one to reach 200 moves. Since each move involves a step by each player, there are at most 400 positions involved in most games. If we traversed the tree of chess positions only to that depth, we would only need enough memory to store the 400 positions on a single path at a time. This much memory is easily available on the smallest computers you are likely to use.

So perfect chess playing is a problem in PSPACE. (Actually one must be more careful in definitions. There is only a finite number of positions in chess, so in principle you could write down the solution in constant time. But that constant would be very large. Generalized versions of chess on larger boards are in PSPACE.)


## Take out the Theory, Solving integer programs is an NP-hard problem

# Integer Programming
- integer program is exactly the same as a linear program, with the additional const


## Why Must be Integer?
- One example is problems involving the production of large indivisible items, such as airplanes or cars. 
- It usually does not make sense to use a continuous variable to represent the number of airplanes to produce
- Another example of where one would use integer variables is to model a particular state, such as on or off. 
- For example, a unit commitment problem where integer variables are used to represent the state of a particular unit being either on or off.
- Planning of investments also requires integer variables, for example a variable that takes a value of 1 to invest in a warehouse, and 0 to ignore it.
- Finally, integer variables are often used to model logic between different decision, for example that a given tax break is only applicable if a certain investment is made.

## Modeling techniques with integer and binary variables

- Integer and binary variables are very useful to express logical constraints. 
- Here are a few examples of such constraints.

### Indicator variables

- Indicator variables are binary variables used to indicate whether a certain set of conditions is valid (with the variable equal to 1) or not (with the variable equal to 0). 
- For example, consider a production problem where you want to distinguish between two states, namely production above a minimum threshold, and no production. 

To model this, define a binary variable $y$ to take a value of 1 if the production is above the minimum threshold (called minProd), and 0 if there is no production. Assume $production$ is a continuous variable containing the produced quantity. 
This leads to these two constraints.

$$
production \ge minProd * y\\
production \le maxProd * y
$$

Here, maxProd is an upper bound on the production quantity. Thus, if y = 1, the minimum and maximum production bounds hold, and if y = 0, the production is set to zero. 


### Logical constraints - an example

For example, consider an investment decision involving a production plant and two warehouses. 

- If the production plant is invested in, then either warehouse 1 or warehouse 2 may be invested in (not both).

- If the production plant is not invested in, then neither of the two warehouses may be invested in.

Let $yPlant$ be 1 if you decide to invest in the production plant, and 0 otherwise. 
Similar for $yWarehouse1$ and $yWarehouse2$.
Then this example can be modeled as follows:

$$
yWarehouse1 + yWarehouse2 <= yPlant
$$

If $yPlant$ is 0 then both $yWarehouse1$ and $yWarehouse2$ are set to zero. 

On the opposite, if one warehouse variable is set to 1, then $yPlant$ is also set to 1. Finally, this constraint also states that warehouse variables cannot be both equal to 1.


## IP versus MIP

- When all the decision variables in a linear model should take integer values, the model is an Integer Program (or **IP**). 

- When some of the decision variables may also take continuous values, the model is a Mixed Integer Program (or **MIP**). 
- IPs and MIPs are generally much more difficult to solve than LPs.
- The solution complexity increases with the number of possible combinations of the integer variables, and such problems are often referred to as being “combinatorial”.


### MIP
E.g.  supply chain applications where investment decisions may be represented by integers and production quantities are represented by continuous variables.





<p>
<ul>
<img src = "https://ibmdecisionoptimization.github.io/tutorials/jupyter/training/1_39.png?raw=true" >
</ul> 



- This graphic shows the new feasible region where the dots indicate the feasible solutions.  
- That is, solutions where the variables take only integer values.  

## What you should take away from this graphic, is that the feasible region is now a collection of points, as opposed to a solid area.  


- Because, in this example, the integer solution does not lie on an extreme point of the continuous feasible region, LP techniques would not find the integer optimum.  
- To find the integer optimum, you should use an integer programming technique. 


# CPLEX Engine
- In this notebook we are using IBM ILOG DOcplex to formulate the mathematical model in Python
- Use the DOcplex Python library to write the mathematical model in Python. 
- This is done in four steps:

    1. create a instance of docplex.mp.Model to hold all model objects
    2. create decision variables,
    3. create linear constraints,
    4. finally, define the objective.


### But first, we have to import the class Model from the docplex module.

### Step 1: Download the library

First install *docplex* if needed.

In [2]:
import sys
try:
    import docplex.mp
except:
    if hasattr(sys, 'real_prefix'):
        #we are in a virtual env.
        !pip install docplex
    else:
        !pip install --user docplex



### Step 2: Set up the prescriptive engine

- Subscribe to our private cloud offer or Decision Optimization on Cloud solve service here if you do not want to use a local solver.
- Get the service URL and your personal API key and enter your credentials here if accurate:



In [3]:
url = 
key = 


### Step 3: Set up the prescriptive model
Create the model

All objects of the model belong to one model instance.

In [5]:
from docplex.mp.model import Model

### Declaring integer decision variables in DOcplex

DOcplex has specific methods to create integer and binary variables.

In [6]:
im = Model(name='integer_programming')
b = im.binary_var(name='boolean_var')
ijk = im.integer_var(name='int_var')
im.print_information()

Model: integer_programming
 - number of variables: 2
   - binary=1, integer=1, continuous=0
 - number of constraints: 0
   - linear=0
 - parameters: defaults


# Example: a production problem

## Problem description: telephone production

A telephone company produces and sells two kinds of telephones, namely **desk phones and cellular phones**. 

Each type of phone is assembled and painted by the company. 

The objective is to maximize profit

the company has to produce at least 100 of each type of phone.

There are limits in terms of the company’s production capacity, and the company has to calculate the optimal number of each type of phone to produce, 

while not exceeding the capacity of the plant.



## Writing a mathematical model

Convert the descriptive model into a mathematical model:
- Use the two decision variables DeskProduction and CellProduction
- Use the data given in the problem description (remember to convert minutes to hours where appropriate)
- Write the objective as a mathematical expression
- Write the constraints as mathematical expressions (use “=”, “<=”, or “>=”, and name the constraints to describe their purpose)
- Define the domain for the decision variables

### Telephone production: a mathematical model

To express the last two constraints, we model assembly time and painting time as linear combinations of the two productions, resulting in the following mathematical model:

$
maximize:\\
\ \ 12\ desk\_production + 20\ cell\_production\\
subject\ to: \\
\ \   desk\_production >= 100 \\
\ \   cell\_production >= 100 \\
\ \   0.2\ desk\_production + 0.4\ cell\_production <= 400 \\
\ \   0.5\ desk\_production + 0.4\ cell\_production <= 490 \\
$


#### Define the decision variables
- The continuous variable `desk` represents the production of desk telephones.
- The continuous variable `cell` represents the production of cell phones.

In [17]:

# create one model instance, with a name
m = Model(name='telephone_production')


# by default, all variables in Docplex have a lower bound of 0 and infinite upper bound
desk = m.continuous_var(name='desk')
cell = m.continuous_var(name='cell')

#### Set up the constraints
- Desk and cel phone must both be greater than 100
- Assembly time is limited
- Painting time is limited.

In [18]:
# write constraints
# constraint #1: desk production is greater than 100
m.add_constraint(desk >= 100)

# constraint #2: cell production is greater than 100
m.add_constraint(cell >= 100)

# constraint #3: assembly time limit
ct_assembly = m.add_constraint( 0.2 * desk + 0.4 * cell <= 400)

# constraint #4: paiting time limit
ct_painting = m.add_constraint( 0.5 * desk + 0.4 * cell <= 490)

#### Express the objective

We want to maximize the expected revenue.

In [19]:
m.maximize(12 * desk + 20 * cell)

A few remarks about how we formulated the mathemtical model in Python using DOcplex:
- all arithmetic operations (+, \*, \-) are done using Python operators
- comparison operators used in writing linear constraint use Python comparison operators too.

#### Print information about the model

We can print information about the model to see how many objects of each type it holds:

In [20]:
m.print_information()

Model: telephone_production
 - number of variables: 2
   - binary=0, integer=0, continuous=2
 - number of constraints: 4
   - linear=4
 - parameters: defaults


### Graphical representation of a Linear Problem

A simple 2-dimensional LP (with 2 decision variables) can be represented graphically using a x- and y-axis. 

This is often done to demonstrate optimization concepts. 

To do this, follow these steps:
- Assign one variable to the x-axis and the other to the y-axis.
- Draw each of the constraints as you would draw any line in 2 dimensions.
- Use the signs of the constraints (=, <= or >=) to determine which side of each line falls within the feasible region (allowable solutions).
- Draw the objective function as you would draw any line in 2 dimensions, by substituting any value for the objective (for example, 12 * DeskProduction + 20 * CellProduction = 4000) 


#### Feasible set of solutions

<p>
<ul>
<img src = "https://ibmdecisionoptimization.github.io/tutorials/jupyter/training/19.png?raw=true" >
</ul> 

This graphic shows the feasible region for the telephone problem.  
Recall that the feasible region of an LP is the region delimited by the constraints, and it represents all feasible solutions. In this graphic, the variables DeskProduction and CellProduction are abbreviated to be desk and cell instead.  Look at this diagram and search intuitively for the optimal solution.  That is, which combination of desk and cell phones will yield the highest profit.  


#### The optimal solution

<p>
<ul>
<img src = "https://ibmdecisionoptimization.github.io/tutorials/jupyter/training/20.png?raw=true" >
</ul> 

 To find the optimal solution to the LP, you must find values for the decision variables, within the feasible region, that maximize profit as defined by the objective function. In this problem, the objective function is to maximize 
 $$12 * desk + 20 * cell
 $$

To do this, first draw a line representing the objective by substituting a value for the objective.  

Next move the line up (because this is a maximization problem) to find the point where the line last touches the feasible region.  Note that all the solutions on one objective line, such as AB, yield the same objective value.  Other values of the objective will be found along parallel lines (such as line CD). 

In a profit maximizing problem such as this one, these parallel lines are often called isoprofit lines, because all the points along such a line represent the same profit. In a cost minimization problem, they are known as isocost lines. Since all isoprofit lines have the same slope, you can find all other isoprofit lines by pushing the objective value further out, moving in parallel, until the isoprofit lines no longer intersect the feasible region. The last isoprofit line that touches the feasible region defines the largest (therefore maximum) possible value of the objective function. In the case of the telephone production problem, this is found along line EF. 

The optimal solution of a linear program always belongs to an extreme point of the feasible region (that is, at a vertex or an edge).


### Solve with the Decision Optimization solve service

If url and key are None, the Modeling layer will look for a local runtime, otherwise will use the credentials.

Look at the documentation for a good understanding of the various solving/generation modes.

If you're using a Community Edition of CPLEX runtimes, depending on the size of the problem, the solve stage may fail and will need a paying subscription or product installation.

In any case, `Model.solve()` returns a solution object in Python, containing the optimal values of decision variables, if the solve succeeds, or else it returns `None`.

In [21]:
s = m.solve(url=url, key=key)
m.print_solution()

objective: 20600.000
  desk=300.000
  cell=850.000


#### In this case, CPLEX has found an optimal solution at (300, 850). You can check that this point is indeed an extreme point of the feasible region.

## Let's turn  this example into Integer Programming problem

In real life, it is possible that the solution becomes non-integer under certain circumstances, for example:

- Change the availability of the assembly machine to 401 hours
- Change the painting machine availability to 492 hours
- Change the profit for a desk phone to 12.4
- Change the profit for a cell phone to 20.2

The fractional values for profit are quite realistic. Even though the fractional times for availability are not entirely realistic, these are used here to illustrate how fractional solutions may occur. 

Let's solve again the telephone production problem with these new data

In [15]:
lm = Model(name='lp_telephone_production')
desk = lm.continuous_var(name='desk')
cell = lm.continuous_var(name='cell')
# write constraints
# constraint #1: desk production is greater than 100
lm.add_constraint(desk >= 100)

# constraint #2: cell production is greater than 100
lm.add_constraint(cell >= 100)

# constraint #3: assembly time limit
ct_assembly = lm.add_constraint( 0.2 * desk + 0.4 * cell <= 401)

# constraint #4: paiting time limit
ct_painting = lm.add_constraint( 0.5 * desk + 0.4 * cell <= 492)
lm.maximize(12.4 * desk + 20.2 * cell)

ls = lm.solve(url=url, key=key)
lm.print_solution()

objective: 20948.167
  desk=303.333
  cell=850.833


As we can see the optimal solution contains fractional values for number of telephones, which are not realistic.
To ensure we get integer values in the solution, we can use integer decision variables.

Let's solve a new model, identical except that its two decision variables are declared as _integer_ variables.

In [16]:
im = Model(name='ip_telephone_production')
desk = im.integer_var(name='desk')
cell = im.integer_var(name='cell')
# write constraints
# constraint #1: desk production is greater than 100
im.add_constraint(desk >= 100)

# constraint #2: cell production is greater than 100
im.add_constraint(cell >= 100)

# constraint #3: assembly time limit
im.add_constraint( 0.2 * desk + 0.4 * cell <= 401)

# constraint #4: paiting time limit
im.add_constraint( 0.5 * desk + 0.4 * cell <= 492)
im.maximize(12.4 * desk + 20.2 * cell)

si = im.solve(url=url, key=key)
im.print_solution()

objective: 20947.400
  desk=303
  cell=851


# LP Relaxation Method
- When we solve an integer programming problem, one common approach is to relax all the integer variables to become continuous. 
- Then we will have a linear programming problem which is easy to solve (This version of the problem is called the linear programming relaxation of your original integer program).
- The solution that you obtain from solving the LP relaxation will most likely not satisfy the integrality constraints present in your original problem. 
- To remedy this, two approaches are introduced:

## 1. Cutting plane methods. 
These can also be used to solve convex problems with integer constraints.

## 2. Branch and bound methods. 
These can also be used to solve nonliear problems with integer constraints (MINLP).


# Cutting Plane Method (Gomory Cut)

- When we form the LP relaxation, we relax the problem in the sense that the feasible domain becomes larger. 
- Gomory cuts try to find and append inequalities that chop off parts of the new feasible region to our LP relaxation tableu, 
- removing fractional solutions until we stumble upon an integer solution. They will not, however, remove any integer feasible points. In a sense, they chop off excess regions of the feasible domain (like cutting off the crust of a toast).

- Namely, we generate valid inequalities (called cuts) from the LP relaxation tableu, 
- that are violated at the current LP fractional solution; meaning that we need to resolve to recover feasibility. 
- We resolve the problem and if we get another fractional solution, we repeat the process. 
- We stop whenever we reach a solution that satisfies all the integrality constraints of the original integer programming problem.



<p>
<ul>
<img src = "https://qph.ec.quoracdn.net/main-qimg-81a7af9b262e29ca1c5cf3088d126243-c" >
</ul> 


- The fractional (LP solution) is at the intersection of the two constraints 
- whereas our integer feasible solutions are the solid dots. 
- The dashed line is our cut. Note how it cuts off the current fractional solution and doesn't cut off any integer feasible points.

Details: https://en.wikipedia.org/wiki/Cutting-plane_method


##  Branch and Bound method

- The _branch and bound_ method, implemented in CPLEX Mixed-Integer Optimizer, provides an efficient way to solve IP and MIP problems. 

- This method begins by relaxing the integer requirement and treating the problem as an LP. 
- If all the variables take integer values, the solution is complete. 
- If not, the algorithm begins a tree search.  You’ll now see an example of this tree search. 

**Consider this integer programming problem, involving**
- an objective to maximize, 
- three constraints, 
- three non-negative integer variables 


$
maximize\ x + y + 2 z\\
subject\ to: 7x + 2y + 3z <= 36\\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 5x + 4y + 7z <= 42\\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ 2x + 3y + 5z <= 28\\
\ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ \ x,y,z \ge 0
$

### Branch and Bound: the root node

- The first node of the branch and bound tree is the LP relaxation of the original IP model.  
- LP relaxation means that the integer variables have been relaxed to be continuous variables.  
- The solution to the LP relaxation of a maximization IP, such as this, provides an upper bound to the original problem.
- In this case that bound is eleven and five elevenths.  
- The current lower bound is minus infinity. 

<p>
<ul>
<img src = "https://ibmdecisionoptimization.github.io/tutorials/jupyter/training/1_43.png?raw=true" >
</ul> 

#### In this case, the solution is fractional and the tree search continues in order to try and find an integer solution.

### Branch and Bound: branching on a variable

- The algorithm next chooses one of the variables to branch on, in this case $x$, and adds two constraints to create two subproblems.  
- These two constraints are based on the relaxed value of x, namely one and three elevenths.  
- In the one subproblem, $x$ is required to be less than or equal to one, and in the other problem, 
- $x$ is required to be greater than or equal to two, in order to eliminate the fractional solution found.  
- IP2 gives another fractional solution, but IP3 gives an integer solution.  
- This integer solution of 10 is the new lower bound to the original maximization problem, 
- because it is the best current solution to the maximization problem.  

<p>
<ul>
<img src = "https://ibmdecisionoptimization.github.io/tutorials/jupyter/training/1_44.png?raw=true" >
</ul> 

- The algorithm will terminate when the gap between the upper and lower bounds is sufficiently small, 
- but at this point there is still more of the tree to explore.

### Branch and Bound: iteration

'

- Next, two more subproblems are created from IP4, namely one with y less than or equal to zero in IP6, 
- and one with y greater than or equal to 1 in IP5.  
- IP6 yields an integer solution of 11, which is an improvement of the previously found lower bound of 10. 
- IP5 gives a fractional solution and can be explored further.

<p>
<ul>
<img src = "https://ibmdecisionoptimization.github.io/tutorials/jupyter/training/1_47.png?raw=true" >
</ul> 

- So another two subproblems are created from IP5, namely IP8 with z less than or equal to 4, and IP7 with z greater than or equal to 5.  
- However, the constraint added for IP4 specifies that z must be less than or equal to 5, so node IP7 immediately yields an integer solution with an objective value of 11, which is the same objective as for IP6.  
- IP8 yields an integer solution with objective value of 9, which is a worse solution than those previously found and IP8 can therefore be discarded.  

### The optimal solution reported is the integer solution with the best objective value that was found first, namely the solution to IP6.

## CPLEX with BBM
- The progess of the Branch & Bound algorithm can be monitored by looking at the CPLEX the _log_. 
- Adding the keyword argument `log_output=True` to the `Model.solve()` method will print the log on the standard output.
- You can see the best bound going down until the gap closes and the final solution of 11 is returned.


In [7]:
bbm = Model(name='b&b')
x, y, z = bbm.integer_var_list(3, name=['x', 'y', 'z'])
bbm.maximize(x + y + 2*z)
bbm.add_constraint(7*x + 2*y + 3*z <= 36)
bbm.add_constraint(5*x + 4*y + 7*z <= 42)
bbm.add_constraint(2*x + 3*y + 5*z <= 28)
bbm.solve(url=url, key=key,log_output=True);

[2018-02-09T06:09:02Z, INFO] CPLEX version 12070100
[2018-02-09T06:09:02Z, INFO] Parameter file:
[2018-02-09T06:09:02Z, INFO] # -- This content is generated by DOcplex
[2018-02-09T06:09:02Z, INFO] CPLEX Parameter File Version 12.6.3.0
[2018-02-09T06:09:02Z, INFO] # --- end of generated prm data ---
[2018-02-09T06:09:02Z, WARN] Warning:  CPLEX version found in file file.prm (12.6.3.0) does not match
[2018-02-09T06:09:02Z, WARN]           this CPLEX version (12.7.1.0).
[2018-02-09T06:09:02Z, WARN] Changed parameter CPX_PARAM_THREADS from 0 to 10
[2018-02-09T06:09:02Z, INFO] Param[1,067] = 10
[2018-02-09T06:09:02Z, INFO] Param[1,130] = utf-8
[2018-02-09T06:09:02Z, INFO] Param[1,132] = -1
[2018-02-09T06:09:02Z, INFO] CPXPARAM_WorkDir                                 "/var/job_processor/work/tmp/java/job8413213856012148299.tmp"
[2018-02-09T06:09:02Z, INFO] CPXPARAM_Threads                                 10
[2018-02-09T06:09:02Z, INFO] CPXPARAM_Read_APIEncoding                        "utf-8"

## Modeling yes/no decisions with binary variables: an example

### Binary variables are often used to model yes/no decisions.  

Consider a telephone production problem: The company is considering replacing the assembly machine with a newer machine that requires less time for cell phones, namely 18 minutes per phone, but more time for desk phones, namely 15 minutes per phone. This machine is available for 430 hours, as opposed to the 400 hours of the existing assembly machine, because it requires less downtime. 

We will design and write a model that uses binary variables to help the company choose between the two machines. 

The steps to formulate the mixed-integer model are:
- Add four new variables (desk1, desk2, cell1, and cell2, to indicate the production on assembly machines 1 and 2, respectively.
- Add two constraints to define the total production of desk and cell to equal the sum of production from the two assembly machines.
- Rewrite the constraint for assembly machine 1 to use the new variables for that machine (desk1 and cell1).
- Add a similar constraint for the production on assembly machine 2.
- Define a Boolean variable, y, to take a value of 1 if assembly machine 1 is chosen, and 0 if assembly machine 2 is chosen.
- Use the y variable to set the production to zero for the machine that is not chosen.


### Implementing the yes/no decision model with DOcplex

First, create a model instance.

In [10]:
tm2 = Model('decision_phone')

#### Setup decision variables

we create two sets of (desk, cell) integer variables, one per machine type, plus the total production variables.
Note that the total production variables do not need to be declared if the four typed productions are integers.
As the sum of two integers, they will always be integers; the less we have of integer variables, the easier CPLEX willsolve the problem.

In addition, we define an extra binary variable $z$ to model the choice we are facing: use machine #1 or machine #2.

In [11]:
# variables for total production
desk = tm2.integer_var(name='desk', lb=100)
cell = tm2.continuous_var(name='cell', lb=100)

# two variables per machine type:
desk1 = tm2.integer_var(name='desk1')
cell1 = tm2.integer_var(name='cell1')

desk2 = tm2.integer_var(name='desk2')
cell2 = tm2.integer_var(name='cell2')

# yes no variable
z = tm2.binary_var(name='z')

#### Setup constraints

- The constraint for paiting machine limit is identical to the basic telephone model
- Two extra constraints express the total production as the sum of productions on the two assembly machines.
- Each assembly machine type has its own constraint, in which variable $z$ expresses the exclusive choice between the two.

In [12]:

# total production is sum of type1 + type 2
tm2.add_constraint(desk == desk1 + desk2)
tm2.add_constraint(cell == cell1 + cell2)

# production on assembly machine of type 1 must be less than 400 if y is 1, else 0
tm2.add_constraint(0.2 * desk1 + 0.4 * cell1 <= 400 * z)
# production on assembly machine of type 2 must be less than 430 if y is 0, else 0
tm2.add_constraint(0.25 * desk2 + 0.3 * cell2 <= 430 * (1-z))

# painting machine limit is identical
# constraint #4: paiting time limit
tm2.add_constraint( 0.5 * desk + 0.4 * cell <= 490)

tm2.print_information()

Model: decision_phone
 - number of variables: 7
   - binary=1, integer=5, continuous=1
 - number of constraints: 5
   - linear=5
 - parameters: defaults


#### Expressing the objective

The objective is identical: maximize total profit, using total productions.

In [13]:
tm2.maximize(12 * desk + 20 * cell)

#### Solve with the Decision Optimization solve service

In [14]:
tm2s= tm2.solve(url=url, key=key,log_output=True)
assert tm2s
tm2.print_solution()

[2018-02-09T06:12:10Z, INFO] CPLEX version 12070100
[2018-02-09T06:12:10Z, INFO] Parameter file:
[2018-02-09T06:12:10Z, INFO] # -- This content is generated by DOcplex
[2018-02-09T06:12:10Z, INFO] CPLEX Parameter File Version 12.6.3.0
[2018-02-09T06:12:10Z, INFO] # --- end of generated prm data ---
[2018-02-09T06:12:10Z, WARN] Warning:  CPLEX version found in file file.prm (12.6.3.0) does not match
[2018-02-09T06:12:10Z, WARN]           this CPLEX version (12.7.1.0).
[2018-02-09T06:12:10Z, WARN] Changed parameter CPX_PARAM_THREADS from 0 to 10
[2018-02-09T06:12:10Z, INFO] Param[1,067] = 10
[2018-02-09T06:12:10Z, INFO] Param[1,130] = utf-8
[2018-02-09T06:12:10Z, INFO] Param[1,132] = -1
[2018-02-09T06:12:10Z, INFO] CPXPARAM_WorkDir                                 "/var/job_processor/work/tmp/java/job3105977193318409067.tmp"
[2018-02-09T06:12:10Z, INFO] CPXPARAM_Threads                                 10
[2018-02-09T06:12:10Z, INFO] CPXPARAM_Read_APIEncoding                        "utf-8"

#### Conclusion

This model demonstrates that the optimal solution is to use machine #2 , producing 100 deskphones and 1100 cell phones.

### Using binary variables for logical decisions

What if the company had to choose between 3 possible candidates for the assembly machine, as opposed to two?

The above model can be generalized with three binary variables $z1$, $z2$, $z3$ each of which is equal to 1 only if machine type 1,2, or 3 is used. But then we need to express that _exactly_ one of those variables must be equal to 1. How can we achive this?

The answer is to add  the following constraint to the model:

$$
z_{1} + z_{2} + z_{3} = 1
$$

Thus, if one of zs variables is equal to 0, the two other areequal to zero (remember binary variables can take value 0 or 1).